src/main/java/io/stargate/sgv2/jsonapi/service/operation/reranking/ScoredDocument.java

```diff
@@ -38,16 +56,101 @@ private static ObjectNode checkCast(JsonNode document) {
           "id must be a value node, got id.getNodeType()=%s".formatted(id.getNodeType()));
     }
 
-    this.passage = Objects.requireNonNull(passage, "passage cannot be null");
+    if (passage != null && passage.isBlank()) {
+      throw new IllegalArgumentException("passage cannot be blank");
+    }
+
+    this.passage = Optional.ofNullable(passage);
     this.document = Objects.requireNonNull(document, "document cannot be null");
     this.scores = Objects.requireNonNull(scores, "scores cannot be null");
   }
 
+  /**
+   * Factory to create ane instance based on a document.
+   *
+   * @param rank Rank of this document, the order it appeared in the result set.
+   * @param rankSource The source of the rank
+   * @param document Document to extract the id and passage from.
+   * @param passageLocator Locator to find the passage in the document.
+   * @return A new instance of ScoredDocument.
+   */
+  static ScoredDocument create(
+      int rank, Rank.RankSource rankSource, JsonNode document, PathMatchLocator passageLocator) {
+
+    Objects.requireNonNull(document, "document cannot be null");
+    Objects.requireNonNull(passageLocator, "passageLocator cannot be null");
+
+    var documentId =
+        switch (document.get(DOC_ID)) {
+          case null -> throw new IllegalArgumentException("Document id cannot be missing");
+          case NullNode ignored ->
+              throw new IllegalArgumentException("Document id cannot be a NullNode");
+          case JsonNode jsonNode -> jsonNode;
+        };
+
+    DocumentScores vectorScores;
+    var similarityField = document.get(VECTOR_FUNCTION_SIMILARITY_FIELD);
+    if (rankSource == Rank.RankSource.VECTOR) {
+      if (similarityField instanceof NumericNode numberNode) {
+        vectorScores = DocumentScores.fromVectorRead(numberNode.floatValue(), rank);
+      } else {
+        throw new IllegalArgumentException(
+            "%s document field is not a number or is missing for doc _id %s"
```

Comment:
there's JsonUtil method `nodeTypeAsString(node)` for "pretty" type description fwtw ("Object", "Array", "Number", "<null>")

Reply:
changed 

```java
        throw new IllegalArgumentException(
            "%s document field is not a number or is missing type=%s _id=%s"
                .formatted(VECTOR_FUNCTION_SIMILARITY_FIELD, nodeTypeAsString(similarityField), documentId.asText()));
      }
```